py/dml/c_backend.py

```diff
@@ -1958,62 +1956,104 @@ def scramble_argname(name):
 
 def trait_trampoline_name(method, vtable_trait):
     return "%s__trampoline_from_%s" % (
-        crep.cref(method), vtable_trait.name)
+        crep.cref_method(method), vtable_trait.name)
 
 def flatten_object_subtree(node):
     '''return a list of all composite subobjects inside node'''
     return [node] + node.get_recursive_components(
         'event', 'port', 'implement', 'attribute', 'connect',
         'interface', 'bank', 'group', 'register', 'field')
 
-def init_trait_vtables_for_node(node, param_values, dims, parent_indices):
-    init_calls = []
-    if node.isindexed():
-        if all(not o.traits for o in flatten_object_subtree(node)):
-            return
-        indices = ()
-        for (arridx, arrlen) in enumerate(node.arraylens()):
-            idx = "_i%d" % (arridx + node.nonlocal_dimensions())
-            dims = dims + (arrlen,)
-            indices += (mkLit(node.site, idx, TInt(32, True)),)
-        indices = parent_indices + indices
-    else:
-        indices = parent_indices
-    if node.traits:
-        index_str = ''.join('[%s]' % (i,) for i in indices)
-        method_overrides = node.traits.method_overrides
-        param_overrides = param_values[node]
-        for trait in node.traits.referenced:
-            args = []
-            for name in tinit_args(trait):
-                member_kind = trait.member_kind(name)
-                if member_kind == 'method':
+def init_trait_vtable(node, trait, param_overrides):
```

Comment:
This method is somewhat confusing to read and grok, in particular, as the code populating `param_init` and `param_decl` is so large, and comes first, it was not immediately clear to me to me that the elements of `param_init` would be emitted inside a for loop, and my first impression was that the code was bugged and only initialized a single element of the array. Perhaps splitting this function would make it easier to read and understand.

Reply:
i'll add comment that describes args/param_decl/param_init